File: apps/macos-menubar/DeepCode/Sidecar/ProtocolTypes.swift
import Foundation

// MARK: - Shared types

struct SkillInfo: Codable {
    let name: String
    let path: String
    let description: String
    let isLoaded: Bool?
}

struct SlashCommandItem: Decodable {
    let kind: String          // "skill" | "skills" | "new" | "resume" | "exit"
    let name: String
    let label: String
    let description: String
    let skill: SkillInfo?
}

// MARK: - Inbound (App → CLI)

enum ClientCommand: Encodable {
    case submit(id: String, text: String, imageUrls: [String]? = nil, skills: [SkillInfo]? = nil)
    case interrupt(id: String?)
    case listSessions(id: String)
    case loadSession(id: String, sessionId: String)
    case newSession(id: String)
    case changeProjectRoot(id: String, path: String)
    case listSlashCommands(id: String)
    case readClipboardImage(id: String)

    private enum CodingKeys: String, CodingKey {
        case type, id, text, sessionId, path
        case imageUrls, skills
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .submit(id, text, imageUrls, skills):
            try container.encode("submit", forKey: .type)
            try container.encode(id, forKey: .id)
            try container.encode(text, forKey: .text)
            if let imageUrls = imageUrls, !imageUrls.isEmpty {
                try container.encode(imageUrls, forKey: .imageUrls)
            }
            if let skills = skills, !skills.isEmpty {
                try container.encode(skills, forKey: .skills)
            }
        case let .interrupt(id):
            try container.encode("interrupt", forKey: .type)
            if let id = id { try container.encode(id, forKey: .id) }
        case let .listSessions(id):
            try container.encode("list_sessions", forKey: .type)
            try container.encode(id, forKey: .id)
        case let .loadSession(id, sessionId):
            try container.encode("load_session", forKey: .type)
            try container.encode(id, forKey: .id)
            try container.encode(sessionId, forKey: .sessionId)
        case let .newSession(id):
            try container.encode("new_session", forKey: .type)
            try container.encode(id, forKey: .id)
        case let .changeProjectRoot(id, path):
            try container.encode("change_project_root", forKey: .type)
            try container.encode(id, forKey: .id)
            try container.encode(path, forKey: .path)
        case let .listSlashCommands(id):
            try container.encode("list_slash_commands", forKey: .type)
            try container.encode(id, forKey: .id)
        case let .readClipboardImage(id):
            try container.encode("read_clipboard_image", forKey: .type)
            try container.encode(id, forKey: .id)
        }
    }
}

// MARK: - Outbound (CLI → App)

struct ServerSessionEntry: Decodable {
    let id: String
    let summary: String?
    let assistantReply: String?
    let assistantThinking: String?
    let assistantRefusal: String?
    let status: String
    let failReason: String?
    let activeTokens: Int?
    let createTime: String
    let updateTime: String
}

struct ServerMessage: Decodable {
    let id: String
    let sessionId: String
    let role: String
    let content: String?
    let visible: Bool?
    let createTime: String?
    let meta: ServerMessageMeta?
    let messageParams: ServerMessageParams?
}

struct ServerMessageMeta: Decodable {
    let asThinking: Bool?
    let isSummary: Bool?
    let paramsMd: String?
    let resultMd: String?
    let function: ServerFunctionRef?
}

/// CLI sometimes returns the assistant reply only inside `messageParams.reasoning_content`
/// (thinking-only models like deepseek-v3.2 on Volcano Ark). UI must fall back to it
/// when `content` is empty.
struct ServerMessageParams: Decodable {
    let reasoningContent: String?

    private enum CodingKeys: String, CodingKey {
        case reasoningContent = "reasoning_content"
    }
}

struct ServerFunctionRef: Decodable {
    let name: String?
    let arguments: String?
}

enum ServerEvent: Decodable {
    case ready(version: String, machineId: String?, projectRoot: String)
    case session(entry: ServerSessionEntry)
    case stream(phase: String, estimatedTokens: Int, formattedTokens: String, sessionId: String?)
    case message(ServerMessage)
    case sessionsList(id: String, sessions: [ServerSessionEntry])
    case sessionLoaded(id: String, sessionId: String, messages: [ServerMessage])
    case error(id: String?, message: String)
    case done(id: String, status: String)
    case ack(id: String)
    case projectRootChanged(id: String, path: String, skills: [SkillInfo])
    case slashCommands(id: String, commands: [SlashCommandItem])
    case clipboardImage(id: String, dataUrl: String?, error: String?)
    case unknown(rawType: String)

    private enum CodingKeys: String, CodingKey {
        case type
        case version, machineId, projectRoot
        case entry
        case phase, estimatedTokens, formattedTokens, sessionId
        case message
        case id, sessions, messages
        case error
        case status
        case path, skills
        case commands
        case dataUrl
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        switch type {
        case "ready":
            let version = try container.decode(String.self, forKey: .version)
            let machineId = try container.decodeIfPresent(String.self, forKey: .machineId)
            let projectRoot = try container.decode(String.self, forKey: .projectRoot)
            self = .ready(version: version, machineId: machineId, projectRoot: projectRoot)
        case "session":
            let entry = try container.decode(ServerSessionEntry.self, forKey: .entry)
            self = .session(entry: entry)
        case "stream":
            let phase = try container.decode(String.self, forKey: .phase)
            let est = (try? container.decode(Int.self, forKey: .estimatedTokens)) ?? 0
            let formatted = (try? container.decode(String.self, forKey: .formattedTokens)) ?? "0"
            let sid = try container.decodeIfPresent(String.self, forKey: .sessionId)
            self = .stream(phase: phase, estimatedTokens: est, formattedTokens: formatted, sessionId: sid)
        case "message":
            let msg = try container.decode(ServerMessage.self, forKey: .message)
            self = .message(msg)
        case "sessions_list":
            let id = try container.decode(String.self, forKey: .id)
            let sessions = try container.decode([ServerSessionEntry].self, forKey: .sessions)
            self = .sessionsList(id: id, sessions: sessions)
        case "session_loaded":
            let id = try container.decode(String.self, forKey: .id)
            let sid = try container.decode(String.self, forKey: .sessionId)
            let messages = try container.decode([ServerMessage].self, forKey: .messages)
            self = .sessionLoaded(id: id, sessionId: sid, messages: messages)
        case "error":
            let id = try container.decodeIfPresent(String.self, forKey: .id)
            let message = try container.decode(String.self, forKey: .error)
            self = .error(id: id, message: message)
        case "done":
            let id = try container.decode(String.self, forKey: .id)
            let status = try container.decode(String.self, forKey: .status)
            self = .done(id: id, status: status)
        case "ack":
            let id = try container.decode(String.self, forKey: .id)
            self = .ack(id: id)
        case "project_root_changed":
            let id = try container.decode(String.self, forKey: .id)
            let path = try container.decode(String.self, forKey: .path)
            let skills = (try? container.decode([SkillInfo].self, forKey: .skills)) ?? []
            self = .projectRootChanged(id: id, path: path, skills: skills)
        case "slash_commands":
            let id = try container.decode(String.self, forKey: .id)
            let commands = (try? container.decode([SlashCommandItem].self, forKey: .commands)) ?? []
            self = .slashCommands(id: id, commands: commands)
        case "clipboard_image":
            let id = try container.decode(String.self, forKey: .id)
            let dataUrl = try container.decodeIfPresent(String.self, forKey: .dataUrl)
            let error = try container.decodeIfPresent(String.self, forKey: .error)
            self = .clipboardImage(id: id, dataUrl: dataUrl, error: error)
        default:
            self = .unknown(rawType: type)
        }
    }
}

File: apps/windows-tray/DeepCode/Sidecar/ProtocolTypes.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace DeepCode.Sidecar;

// --- Shared --------------------------------------------------------------

public sealed record SkillInfo(string Name, string Path, string Description, bool? IsLoaded);

public sealed record SlashCommandItem(
    string Kind,        // "skill" | "skills" | "new" | "resume" | "exit"
    string Name,
    string Label,
    string Description,
    SkillInfo? Skill);

public sealed record ServerSessionEntry(
    string Id,
    string? Summary,
    string? AssistantReply,
    string? AssistantThinking,
    string? AssistantRefusal,
    string Status,
    string? FailReason,
    int? ActiveTokens,
    string CreateTime,
    string UpdateTime);

public sealed record ServerFunctionRef(string? Name, string? Arguments);

public sealed record ServerMessageMeta(
    bool? AsThinking,
    bool? IsSummary,
    string? ParamsMd,
    string? ResultMd,
    ServerFunctionRef? Function);

public sealed record ServerMessageParams(
    [property: JsonPropertyName("reasoning_content")] string? ReasoningContent);

public sealed record ServerMessage(
    string Id,
    string SessionId,
    string Role,
    string? Content,
    bool? Visible,
    string? CreateTime,
    ServerMessageMeta? Meta,
    ServerMessageParams? MessageParams);

// --- Outbound (CLI → App) ------------------------------------------------

public abstract record ServerEvent;
public sealed record ReadyEvent(string Version, string? MachineId, string ProjectRoot) : ServerEvent;
public sealed record SessionEvent(ServerSessionEntry Entry) : ServerEvent;
public sealed record StreamEvent(string Phase, int EstimatedTokens, string FormattedTokens, string? SessionId) : ServerEvent;
public sealed record MessageEvent(ServerMessage Message) : ServerEvent;
public sealed record SessionsListEvent(string Id, IReadOnlyList<ServerSessionEntry> Sessions) : ServerEvent;
public sealed record SessionLoadedEvent(string Id, string SessionId, IReadOnlyList<ServerMessage> Messages) : ServerEvent;
public sealed record ErrorEvent(string? Id, string Message) : ServerEvent;
public sealed record DoneEvent(string Id, string Status) : ServerEvent;
public sealed record AckEvent(string Id) : ServerEvent;
public sealed record ProjectRootChangedEvent(string Id, string Path, IReadOnlyList<SkillInfo> Skills) : ServerEvent;
public sealed record SlashCommandsEvent(string Id, IReadOnlyList<SlashCommandItem> Commands) : ServerEvent;
public sealed record ClipboardImageEvent(string Id, string? DataUrl, string? Error) : ServerEvent;
public sealed record UnknownEvent(string RawType) : ServerEvent;

// --- Inbound (App → CLI) -------------------------------------------------

public abstract record ClientCommand
{
    public abstract JsonObject ToJson();
}

public sealed record SubmitCommand(string Id, string Text, IReadOnlyList<string>? ImageUrls = null, IReadOnlyList<SkillInfo>? Skills = null) : ClientCommand
{
    public override JsonObject ToJson()
    {
        var obj = new JsonObject
        {
            ["type"] = "submit",
            ["id"] = Id,
            ["text"] = Text,
        };
        if (ImageUrls is { Count: > 0 })
        {
            obj["imageUrls"] = JsonNode.Parse(JsonSerializer.Serialize(ImageUrls))!;
        }
        if (Skills is { Count: > 0 })
        {
            obj["skills"] = JsonNode.Parse(JsonSerializer.Serialize(Skills, ProtocolJson.Options))!;
        }
        return obj;
    }
}

public sealed record InterruptCommand(string? Id) : ClientCommand
{
    public override JsonObject ToJson()
    {
        var obj = new JsonObject { ["type"] = "interrupt" };
        if (Id is not null) obj["id"] = Id;
        return obj;
    }
}

public sealed record ListSessionsCommand(string Id) : ClientCommand
{
    public override JsonObject ToJson() => new() { ["type"] = "list_sessions", ["id"] = Id };
}

public sealed record LoadSessionCommand(string Id, string SessionId) : ClientCommand
{
    public override JsonObject ToJson() => new()
    {
        ["type"] = "load_session",
        ["id"] = Id,
        ["sessionId"] = SessionId,
    };
}

public sealed record NewSessionCommand(string Id) : ClientCommand
{
    public override JsonObject ToJson() => new() { ["type"] = "new_session", ["id"] = Id };
}

public sealed record ChangeProjectRootCommand(string Id, string Path) : ClientCommand
{
    public override JsonObject ToJson() => new()
    {
        ["type"] = "change_project_root",
        ["id"] = Id,
        ["path"] = Path,
    };
}

public sealed record ListSlashCommandsCommand(string Id) : ClientCommand
{
    public override JsonObject ToJson() => new() { ["type"] = "list_slash_commands", ["id"] = Id };
}

public sealed record ReadClipboardImageCommand(string Id) : ClientCommand
{
    public override JsonObject ToJson() => new() { ["type"] = "read_clipboard_image", ["id"] = Id };
}

// --- Shared JSON options + manual ServerEvent dispatch -------------------

internal static class ProtocolJson
{
    public static readonly JsonSerializerOptions Options = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        ReadCommentHandling = JsonCommentHandling.Skip,
        AllowTrailingCommas = true,
    };

    public static ServerEvent DecodeEvent(string line)
    {
        JsonNode? root;
        try { root = JsonNode.Parse(line); }
        catch (Exception ex) { return new ErrorEvent(null, $"Decode failed: {ex.Message}. Line: {line}"); }
        if (root is not JsonObject obj) return new ErrorEvent(null, $"Expected object, got: {line}");

        var type = obj["type"]?.GetValue<string>() ?? "";
        try
        {
            return type switch
            {
                "ready" => new ReadyEvent(
                    obj["version"]!.GetValue<string>(),
                    obj["machineId"]?.GetValue<string>(),
                    obj["projectRoot"]!.GetValue<string>()),
                "session" => new SessionEvent(obj["entry"].Deserialize<ServerSessionEntry>(Options)!),
                "stream" => new StreamEvent(
                    obj["phase"]!.GetValue<string>(),
                    obj["estimatedTokens"]?.GetValue<int>() ?? 0,
                    obj["formattedTokens"]?.GetValue<string>() ?? "0",
                    obj["sessionId"]?.GetValue<string>()),
                "message" => new MessageEvent(obj["message"].Deserialize<ServerMessage>(Options)!),
                "sessions_list" => new SessionsListEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["sessions"].Deserialize<List<ServerSessionEntry>>(Options) ?? new()),
                "session_loaded" => new SessionLoadedEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["sessionId"]!.GetValue<string>(),
                    obj["messages"].Deserialize<List<ServerMessage>>(Options) ?? new()),
                "error" => new ErrorEvent(
                    obj["id"]?.GetValue<string>(),
                    obj["error"]?.GetValue<string>() ?? "unknown error"),
                "done" => new DoneEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["status"]!.GetValue<string>()),
                "ack" => new AckEvent(obj["id"]!.GetValue<string>()),
                "project_root_changed" => new ProjectRootChangedEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["path"]!.GetValue<string>(),
                    obj["skills"].Deserialize<List<SkillInfo>>(Options) ?? new()),
                "slash_commands" => new SlashCommandsEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["commands"].Deserialize<List<SlashCommandItem>>(Options) ?? new()),
                "clipboard_image" => new ClipboardImageEvent(
                    obj["id"]!.GetValue<string>(),
                    obj["dataUrl"]?.GetValue<string>(),
                    obj["error"]?.GetValue<string>()),
                _ => new UnknownEvent(type),
            };
        }
        catch (Exception ex)
        {
            return new ErrorEvent(null, $"Decode '{type}' failed: {ex.Message}");
        }
    }
}

internal static class JsonNodeExtensions
{
    public static T? Deserialize<T>(this JsonNode? node, JsonSerializerOptions options)
        => node is null ? default : JsonSerializer.Deserialize<T>(node.ToJsonString(), options);
}
